src/main/java/com/azure/cosmos/kafka/connect/sink/BulkWriter.java

```diff
@@ -0,0 +1,146 @@
+                        bulkOperationResponse.getException(),
+                        bulkOperationResponse.getOperation().getContext());
+
+                if (ExceptionsHelper.canBeTransientFailure(exception)) {
+                    sinkWriteResponse.getFailedRecordResponses().add(new SinkOperationFailedResponse(sinkRecord, exception));
+                } else {
+                    sinkWriteResponse.getFailedRecordResponses().add(0, new SinkOperationFailedResponse(sinkRecord, exception));
+                }
+            } else {
+                sinkWriteResponse.getSucceededRecords().add(sinkRecord);
+            }
+        }
+
+        return sinkWriteResponse;
+    }
+
+    private PartitionKey getPartitionKeyValue(Object recordValue) {
+        checkArgument(recordValue instanceof Map, "Argument 'recordValue' is not valid map format.");
+
+        //TODO: examine the code here for sub-partition
+        String partitionKeyPath = StringUtils.join(this.partitionKeyDefinition.getPaths(), "");
+        Map<String, Object> recordMap = (Map<String, Object>) recordValue;
+        Object partitionKeyValue = recordMap.get(partitionKeyPath.substring(1));
+        PartitionKeyInternal partitionKeyInternal = PartitionKeyInternal.fromObjectArray(Collections.singletonList(partitionKeyValue), false);
+
+        return ImplementationBridgeHelpers
+                .PartitionKeyHelper
+                .getPartitionKeyAccessor()
+                .toPartitionKey(partitionKeyInternal);
+    }
+
+    BulkOperationFailedException handleNonSuccessfulStatusCode(
+            CosmosBulkItemResponse itemResponse,
+            Exception exception,
+            SinkOperationContext sinkOperationContext) {
+
+        int effectiveStatusCode =
+                itemResponse != null
+                        ? itemResponse.getStatusCode()
+                        : (exception != null && exception instanceof CosmosException ? ((CosmosException)exception).getStatusCode() : HttpConstants.StatusCodes.REQUEST_TIMEOUT);
+        int effectiveSubStatusCode =
+                itemResponse != null
+                        ? itemResponse.getSubStatusCode()
+                        : (exception != null && exception instanceof CosmosException ? ((CosmosException)exception).getSubStatusCode() : 0);
+
+        String errorMessage =
+                String.format(
+                        "Request failed with effectiveStatusCode %s, effectiveSubStatusCode %s, kafkaOffset %s, kafkaPartition %s, topic %s",
+                        effectiveStatusCode,
+                        effectiveSubStatusCode,
+                        sinkOperationContext.getKafkaOffset(),
+                        sinkOperationContext.getKafkaPartition(),
+                        sinkOperationContext.getTopic());
+
+
+        return new BulkOperationFailedException(effectiveStatusCode, effectiveSubStatusCode, errorMessage, exception);
+    }
+
+    static class BulkOperationFailedException extends CosmosException {
```

Comment:
should be private?

Reply:
Do we need this class given we already have `CosmosDBWriteException`? 
If so, shouldn't this extend `CosmosDBWriteException` class?

Reply:
I think they serve different purpose. 
BulkOperationFailedException -> internal interpretation of the exceptions we get back from bulkExecutor for each operation. 

CosmosDBWriteException -> it acts like an aggregrated exception which we throw back to kafka framework.